File: python/rocketmq/client/trace_context.py
# -*- coding: utf-8 -*-
"""W3C Trace Context（traceparent）透传（OpenTracing/OTel 场景的消息级上下文）。

格式（W3C Trace Context，https://www.w3.org/TR/trace-context/）::

    traceparent: 00-<trace-id 32hex>-<parent-id 16hex>-<flags 2hex>

* trace-id / parent-id 不能全 0；version 00 时 flags 任意两位 hex；
* 生产侧：消息没有 `traceparent` 属性时注入一个根 span 上下文（opt-in，
  ``enable_trace_context=True``）——Java 客户端把这类注入交给外部链路追踪的
  SendMessageHook（SkyWalking/OTel），本实现内建等价能力，键名沿用 W3C 的小写
  ``traceparent``；已有值**不覆盖**（调用方传播的上下文优先）；
* 消费侧：从 ``MessageExt.properties`` 里取出，供业务做父子 span 关联。
"""

from __future__ import annotations

import os
import secrets
from typing import Optional

from ..common.message import Message, MessageExt

# W3C 键名（小写，与 OTel / HTTP 头一致）
TRACE_CONTEXT_PROPERTY = "traceparent"
TRACE_STATE_PROPERTY = "tracestate"

_ALL_HEX = set("0123456789abcdef")


def _hex(n: int) -> str:
    return secrets.token_hex(n // 2)


def generate_traceparent() -> str:
    """生成合法的根 traceparent：``00-<32hex>-<16hex>-01``（记录采样）。"""
    return "00-%s-%s-01" % (_hex(32), _hex(16))


def is_valid_traceparent(value: Optional[str]) -> bool:
    """按 W3C 语法与"不全 0"规则校验。宽松接受大写 hex（转发不重写）。"""
    if not value:
        return False
    parts = value.strip().split("-")
    if len(parts) != 4:
        return False
    version, trace_id, parent_id, flags = parts
    if version != "00" and not (len(version) == 2 and all(c in _ALL_HEX for c in version.lower())):
        return False
    if version == "ff":
        return False
    if len(trace_id) != 32 or len(parent_id) != 16 or len(flags) != 2:
        return False
    for part, disallow_zero in ((trace_id, True), (parent_id, True), (flags, False)):
        low = part.lower()
        if not all(c in _ALL_HEX for c in low):
            return False
        if disallow_zero and low == "0" * len(low):
            return False
    return True


def child_traceparent(parent: Optional[str]) -> Optional[str]:
    """同一 trace-id 下生成子 span（换 parent-id）；parent 非法返回 None。"""
    if not is_valid_traceparent(parent):
        return None
    parts = parent.strip().split("-")
    return "00-%s-%s-01" % (parts[1].lower(), _hex(16))


def inject_trace_context(message: Message) -> str:
    """消息没有 traceparent 属性时注入根上下文；返回（注入后的）值。

    已有值**不覆盖**——上游传播进来的上下文优先（对齐链路追踪的通用约定）。
    """
    existing = message.get_property(TRACE_CONTEXT_PROPERTY)
    if existing:
        return existing
    tp = generate_traceparent()
    message.put_property(TRACE_CONTEXT_PROPERTY, tp)
    return tp


def extract_traceparent(msg: MessageExt) -> Optional[str]:
    """从消息属性里取出 traceparent（未注入/为空返回 None）。"""
    return msg.get_property(TRACE_CONTEXT_PROPERTY) or None


def trace_context_enabled_from_env() -> bool:
    """env `ROCKETMQ_TRACE_CONTEXT_ENABLE`（对齐其它开关的 env 惯例）。"""
    return os.environ.get("ROCKETMQ_TRACE_CONTEXT_ENABLE", "").strip().lower() in (
        "1", "true", "yes")

File: cpp/src/client/trace_context.cpp
#include "rocketmq/client/trace_context.h"

#include <algorithm>
#include <cctype>
#include <chrono>
#include <cstdlib>
#include <random>
#include <vector>

namespace rocketmq {

namespace {

const char* const kHexDigits = "0123456789abcdef";

std::string randomHex(size_t n) {
    static thread_local std::mt19937_64 rng(
        std::random_device{}() ^ static_cast<uint64_t>(std::chrono::steady_clock::now()
                                                           .time_since_epoch().count()));
    std::string out;
    out.reserve(n);
    for (size_t i = 0; i < n; ++i) {
        out.push_back(kHexDigits[rng() % 16]);
    }
    return out;
}

std::string toLower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
    return s;
}

bool allHex(const std::string& s) {
    if (s.empty()) return false;
    for (char c : s) {
        if (!std::isxdigit(static_cast<unsigned char>(c))) return false;
    }
    return true;
}

bool allZero(const std::string& s) {
    for (char c : s) {
        if (c != '0') return false;
    }
    return true;
}

}  // namespace

std::string generateTraceparent() {
    // 00-<trace-id 32hex>-<parent-id 16hex>-01（flags 01 = recorded）
    return "00-" + randomHex(32) + "-" + randomHex(16) + "-01";
}

bool isValidTraceparent(const std::string& value) {
    if (value.empty()) return false;
    // 宽松：允许首尾空白，段内按小写比对
    std::string v = value;
    // trim
    size_t b = v.find_first_not_of(" \t\r\n");
    size_t e = v.find_last_not_of(" \t\r\n");
    if (b == std::string::npos) return false;
    v = v.substr(b, e - b + 1);

    // 按 '-' 切 4 段
    std::vector<std::string> parts;
    size_t pos = 0;
    while (true) {
        size_t dash = v.find('-', pos);
        if (dash == std::string::npos) {
            parts.push_back(v.substr(pos));
            break;
        }
        parts.push_back(v.substr(pos, dash - pos));
        pos = dash + 1;
    }
    if (parts.size() != 4) return false;
    const std::string version = parts[0];
    const std::string traceId = toLower(parts[1]);
    const std::string parentId = toLower(parts[2]);
    const std::string flags = toLower(parts[3]);
    // version：00 或 2 位 hex 且不是 ff
    if (version == "ff") return false;
    if (version != "00") {
        if (version.size() != 2 || !allHex(version)) return false;
    }
    if (traceId.size() != 32 || !allHex(traceId) || allZero(traceId)) return false;
    if (parentId.size() != 16 || !allHex(parentId) || allZero(parentId)) return false;
    if (flags.size() != 2 || !allHex(flags)) return false;
    return true;
}

std::string childTraceparent(const std::string& parent) {
    if (!isValidTraceparent(parent)) return std::string();
    // 同一 trace-id，换 parent-id（对齐 W3C：子 span 复用 trace-id）
    size_t first = parent.find('-');
    size_t second = parent.find('-', first + 1);
    const std::string traceId = toLower(parent.substr(first + 1, second - first - 1));
    return "00-" + traceId + "-" + randomHex(16) + "-01";
}

std::string injectTraceContext(Message* msg) {
    if (msg == nullptr) return std::string();
    const std::string existing = msg->getProperty(kTraceContextProperty);
    if (!existing.empty()) return existing;  // 上游传播优先，不覆盖
    const std::string tp = generateTraceparent();
    msg->putProperty(kTraceContextProperty, tp);
    return tp;
}

std::string extractTraceparent(const Message& msg) {
    return msg.getProperty(kTraceContextProperty);
}

bool traceContextEnabledFromEnv() {
    const char* v = std::getenv("ROCKETMQ_TRACE_CONTEXT_ENABLE");
    if (v == nullptr) return false;
    std::string s(v);
    for (auto& c : s) c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    return s == "1" || s == "true" || s == "yes";
}

}  // namespace rocketmq
